x/blockdb/database.go

```diff
@@ -0,0 +1,276 @@
+	indexFileName          = "blockdb.idx"
+	dataFileNameFormat     = "blockdb_%d.dat"
+	defaultFilePermissions = 0o666
+
+	// Since 0 is a valid height, math.MaxUint64 is used to indicate unset height.
+	// It is not be possible for block height to be max uint64 as it would overflow the index entry offset
+	unsetHeight = math.MaxUint64
+)
+
+// Database stores blockchain blocks on disk and provides methods to read, and write blocks.
+type Database struct {
+	indexFile *os.File
+	dataDir   string
+	options   DatabaseConfig
+	header    indexFileHeader
+	log       logging.Logger
+	closed    bool
+	fileCache *lru.Cache[int, *os.File]
+
+	// closeMu prevents the database from being closed while in use and prevents
+	// use of a closed database.
+	closeMu sync.RWMutex
+
+	// fileOpenMu prevents race conditions when multiple threads try to open the same data file
+	fileOpenMu sync.Mutex
+
+	// maxBlockHeight tracks the highest block height that has been written to the db, even if there are gaps in the sequence.
+	maxBlockHeight atomic.Uint64
+	// nextDataWriteOffset tracks the next position to write new data in the data file.
+	nextDataWriteOffset atomic.Uint64
+	// maxContiguousHeight tracks the highest block height known to be contiguously stored.
+	maxContiguousHeight atomic.Uint64
+}
+
+func (s *Database) listDataFiles() (map[int]string, int, error) {
+	files, err := os.ReadDir(s.dataDir)
+	if err != nil {
+		return nil, -1, fmt.Errorf("failed to read data directory %s: %w", s.dataDir, err)
+	}
+
+	dataFiles := make(map[int]string)
+	maxIndex := -1
+	for _, file := range files {
+		if file.IsDir() {
+			continue
+		}
+		var index int
+		if n, err := fmt.Sscanf(file.Name(), dataFileNameFormat, &index); n == 1 && err == nil {
+			dataFiles[index] = filepath.Join(s.dataDir, file.Name())
+			if index > maxIndex {
+				maxIndex = index
+			}
+		}
+	}
+
+	return dataFiles, maxIndex, nil
+}
+
+func (s *Database) openAndInitializeIndex(indexDir string, truncate bool) error {
+	indexPath := filepath.Join(indexDir, indexFileName)
+	if err := os.MkdirAll(indexDir, 0o755); err != nil {
+		return fmt.Errorf("failed to create index directory %s: %w", indexDir, err)
+	}
+	openFlags := os.O_RDWR | os.O_CREATE
+	if truncate {
+		openFlags |= os.O_TRUNC
+	}
+	var err error
+	s.indexFile, err = os.OpenFile(indexPath, openFlags, defaultFilePermissions)
+	if err != nil {
+		return fmt.Errorf("failed to open index file %s: %w", indexPath, err)
+	}
+	return s.loadOrInitializeHeader(truncate)
+}
+
+func (s *Database) initializeDataFiles(dataDir string, truncate bool) error {
+	s.dataDir = dataDir
+	if err := os.MkdirAll(dataDir, 0o755); err != nil {
+		return fmt.Errorf("failed to create data directory %s: %w", dataDir, err)
+	}
+
+	if truncate {
+		dataFiles, _, err := s.listDataFiles()
+		if err != nil {
+			return fmt.Errorf("failed to list data files for truncation: %w", err)
+		}
+		for _, filePath := range dataFiles {
+			if err := os.Remove(filePath); err != nil {
```

Comment:
I'm not sure I understand why we're removing data files. Can you explain the rational?